lib/charms/bind/v0/dns_record.py

```diff
@@ -0,0 +1,696 @@
+and provides an `update_relation_data` method to update the relation data by passing
+a `DNSRecordRelationData` data object.
+
+```python
+class DNSRecordProviderCharm(ops.CharmBase):
+    ...
+
+    def _on_config_changed(self, _) -> None:
+        for relation in self.model.relations[self.dns_record.relation_name]:
+            self.dns_record.update_relation_data(relation, self._get_dns_record_data())
+
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "09583c2f9c1d4c0f9a40244cfc20b0c2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 8
+
+PYDEPS = ["pydantic>=2"]
+
+# pylint: disable=wrong-import-position
+import itertools
+import json
+import logging
+from enum import Enum
+from typing import Dict, List, Optional
+from uuid import UUID
+
+import ops
+from pydantic import BaseModel, Field, IPvAnyAddress, ValidationError
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "dns_record"
+
+
+class Status(str, Enum):
+    """Represent the status values.
+
+    Attributes:
+        APPROVED: approved
+        INVALID_CREDENTIALS: invalid_credentials
+        PERMISSION_DENIED: permission_denied
+        CONFLICT: conflict
+        VALIDATION: validation
+        FAILURE: failure
+        UNKNOWN: unknown
+        PENDING: pending
+    """
+
+    APPROVED = "approved"
+    INVALID_CREDENTIALS = "invalid_credentials"
+    PERMISSION_DENIED = "permission_denied"
+    CONFLICT = "conflict"
+    VALIDATION = "validation"
+    FAILURE = "failure"
+    UNKNOWN = "unknown"
+    PENDING = "pending"
+
+
+class RecordType(str, Enum):
+    """Represent the DNS record types.
+
+    Attributes:
+        A: A
+        AAAA: AAAA
+        CNAME: CNAME
+        MX: MX
+        DKIM: DKIM
+        SPF: SPF
+        DMARC: DMARC
+        TXT: TXT
+        CAA: CAA
+        SRV: SRV
+        SVCB: SVCB
+        HTTPS: HTTPS
+        PTR: PTR
+        SOA: SOA
+        NS: NS
+        DS: DS
+        DNSKEY: DNSKEY
+    """
+
+    A = "A"
+    AAAA = "AAAA"
+    CNAME = "CNAME"
+    MX = "MX"
+    DKIM = "DKIM"
+    SPF = "SPF"
+    DMARC = "DMARC"
+    TXT = "TXT"
+    CAA = "CAA"
+    SRV = "SRV"
+    SVCB = "SVCB"
+    HTTPS = "HTTPS"
+    PTR = "PTR"
+    SOA = "SOA"
+    NS = "NS"
+    DS = "DS"
+    DNSKEY = "DNSKEY"
+
+
+class RecordClass(str, Enum):
+    """Represent the DNS record classes.
+
+    Attributes:
+        IN: IN
+    """
+
+    IN = "IN"
+
+
+class DNSProviderData(BaseModel):
+    """Represent the DNS provider data.
+
+    Attributes:
+        uuid: UUID for the domain request.
+        status: status for the domain request.
+        description: status description for the domain request.
+    """
+
+    uuid: UUID
+    status: Status
+    description: Optional[str] = None
+
+    def to_relation_data(self) -> Dict[str, str]:
```

Comment:
Oh what i meant was TypeDict would give type hints for the keys in Dict[str,str] since the relation data shouldn't be too dynamic :)
i.e. a_dict["TYPE_HINTED_KEY"]
but it is really a extra extra quality of life suggestion so up to you, feel free to resolve!